src/mcpwm/operator_config.rs

```diff
@@ -0,0 +1,121 @@
+use esp_idf_sys::mcpwm_operator_config_t__bindgen_ty_1;
+
+use crate::gpio::OutputPin;
+
+use super::{
+    comparator::ComparatorConfig,
+    generator::{GenA, GenB, GeneratorConfig, NoGenCfg, OptionalGenCfg},
+};
+
+#[derive(Default)]
+pub struct OperatorConfig<GENA, GENB> {
+    // TODO: When, how and who should set the flags?
+    // Should that be done automagically, manually or does some specific setting cover all cases?
+    /// Flags for Operator
+    pub(crate) flags: mcpwm_operator_config_t__bindgen_ty_1,
+
+    /// Configuration for Comparator X
+    pub(crate) comparator_x: ComparatorConfig,
+
+    /// Configuration for Comparator Y
+    pub(crate) comparator_y: ComparatorConfig,
+
+    /// Configuration for Generator A
+    pub(crate) generator_a: GENA,
+
+    /// Configuration for Generator B
+    pub(crate) generator_b: GENB,
+}
+
+impl OperatorConfig<NoGenCfg, NoGenCfg> {
+    pub fn default<PA: OutputPin, PB: OutputPin>(
+        pin_a: PA,
+        pin_b: PB,
+    ) -> OperatorConfig<GeneratorConfig<GenA, PA>, GeneratorConfig<GenB, PB>> {
+        OperatorConfig::empty()
+            .cmp_x(Default::default())
+            .cmp_y(Default::default())
+            .gen_a(GeneratorConfig::active_high(pin_a))
+            .gen_b(GeneratorConfig::active_high(pin_b))
```

Comment:
Here for
```rust
.gen_a(GeneratorConfig::active_high(pin_a))
```

due to GeneratorConfig being generic GeneratorChannel, [active_high](https://github.com/esp-rs/esp-idf-hal/blob/691337462ddfb4539197e4a7e3d37b200091d472/src/mcpwm/generator.rs#L161) knows that the generator A should only listen to comparator X.

Similarly for `.gen_b(GeneratorConfig::active_high(pin_b))`, it knows that generator A should only listen for comparator Y.

I am not quite sure how to achieve something similar using enums

Reply:
Do you have any thoughts?

Reply:
The default configuration is to have it work that way, yes. It is also they way it works in IDF 4.4.

However in IDF 5.x you are free to have each generator listen to events from none, any or all comparators.

---

In a future PR when we add support for IDF 4.4 we need to somehow prevent connecting the wrong comparators to the generator.

Reply:
> In a future PR when we add support for IDF 4.4 we need to somehow prevent connecting the wrong comparators to the generator.

This is a bit similar to the [discussion](https://github.com/esp-rs/esp-idf-hal/pull/132#discussion_r1035028499) about how operators and timers are related

Reply:
How about the following simplification then: promoting the connection of the wrong comparator to the generator we can implement as a runtime error, in the name of simplicity? For ESP IDF 4 it will generate a runtime error, and for ESP IDF 5 it will not?

I'm just afraid that we are generating a very heavy API for something that over time will disappear - i.e. ESP IDF 4.4 specifics.

Reply:
Makes, sense. And in the end if this is what you prefer than I will try to do that.

However, I do not think that solves my [initial concern](https://github.com/esp-rs/esp-idf-hal/pull/132#discussion_r1089788561). Further, using the same argument, it would mean that we force users to handle/unwrap an error that in time will not happen. 

Again, not quite happy with the current solution either and I really hear you regarding the enormous amount of generics, but it feels a bit sad to push things we know at compile time to runtime.

Reply:
Well I'm also not happy with pushing things to runtime, especially given that one of the major premises of Rust is compile time safety.

But this is all a fine-balance act, right? As in, I've come to realize that simplicity of use and simpler types is just as important as compile-type safety.

Let me put it another way: if we do this **one** sacrifice - for ESP-IDF 4 only - would that allow us to get rid of most of the generics?

Reply:
>But this is all a fine-balance act, right? As in, I've come to realize that simplicity of use and simpler types is just as important as compile-type safety.

Yes, very much so :)

> Let me put it another way: if we do this one sacrifice - for ESP-IDF 4 only - would that allow us to get rid of most of the generics?

Yes, I think atleast GENA/GENB and CMX/CMPY can be removed if we rethink `OperatorConfig` a bit.

Reply:
I will give it a try :)

Reply:
Thank you so much! I know it's been a long ride...

Reply:
No problem at all, thank *you*!